spec/promoted/ruby/client_spec.rb

```diff
@@ -500,16 +514,61 @@
 
       deliver_resp = client.deliver @input
 
+      # Request insertions should not have properties
       delivery_req[:insertion].each do |insertion|
         expect(insertion.key?(:properties)).to be false
       end
 
+      # But we should have put the properties back on the response insertions.
+      deliver_resp[:insertion].each do |insertion|
+        expect(insertion.key?(:properties)).to be true
+        expect(insertion[:properties].key?(:struct)).to be true
+        expect(insertion[:properties][:struct].key?(:product)).to be true
+      end
+
       expect(deliver_resp).not_to be nil
       expect(deliver_resp.key?(:insertion)).to be true
 
       # No log request generated since there's no experiment and we delivered the request.
       expect(deliver_resp[:log_request]).to be nil
     end
+
+    it "can compact insertions with a subset of properties" do
+      to_compact_delivery_properties_func = Proc.new do |properties|
+          properties[:struct][:product].delete(:url)
```

Comment:
I'll be curious to see what performance looks like eventually for the deep copy approach.  I know deep copy is safer.

Reply:
Yes, per our discussion, I don't think the alternatives are great, unfortunately.